packages/web/components/bridge/immersive/use-bridges-supported-assets.ts

```diff
@@ -0,0 +1,197 @@
+     */
+    const assetAddress_supportedVariants: Record<string, Set<string>> = {};
+
+    /**
+     * Map of supported assets by asset address. This is used to
+     * merge the supported providers for each input asset.
+     */
+    const assetAddress_supportedProviders: Record<string, Set<Bridge>> = {};
+
+    type AssetsByChainId =
+      RouterOutputs["bridgeTransfer"]["getSupportedAssetsByBridge"]["supportedAssets"]["assetsByChainId"];
+
+    const allAssetsByChainId = successfulQueries.reduce((acc, { data }) => {
+      if (!data) return acc;
+
+      // Merge all assets from providers by chain id
+      Object.entries(data.supportedAssets.assetsByChainId).forEach(
+        ([chainId, assets]) => {
+          assets.forEach((asset) => {
+            const { address: rawAddress } = asset;
+            // Use toLowerCase since some providers return addresses in different cases. E.g. Skip and Squid
+            const address = rawAddress.toLowerCase();
+
+            if (!assetAddress_supportedVariants[address]) {
+              assetAddress_supportedVariants[address] = new Set();
+            }
+            if (!assetAddress_supportedProviders[address]) {
+              assetAddress_supportedProviders[address] = new Set<Bridge>();
+            }
+            assetAddress_supportedVariants[address].add(
+              data.supportedAssets.inputAssetAddress
+            );
+            assetAddress_supportedProviders[address].add(
+              data.supportedAssets.providerName
+            );
+          });
+
+          acc[chainId] = acc[chainId] ? [...acc[chainId], ...assets] : assets;
+        }
+      );
+
+      return acc;
+    }, {} as AssetsByChainId);
+
+    const assetEntriesByChainId = Object.entries(allAssetsByChainId).map(
+      ([chainId, assets]) => [
+        chainId,
+        assets
+          .map((asset) => ({
+            ...asset,
+            providerName: undefined,
+            supportedVariants: Array.from(
+              assetAddress_supportedVariants[asset.address.toLowerCase()]
+            ),
+            supportedProviders: Array.from(
+              assetAddress_supportedProviders[asset.address.toLowerCase()]
+            ),
+          }))
+          // Remove Duplicates
+          .filter(
+            (asset, index, originalArray) =>
+              index ===
+              // Use toLowerCase since some providers return addresses in different cases. E.g. Skip and Squid
+              originalArray.findIndex(
+                (t) => t.address.toLowerCase() === asset.address.toLowerCase()
+              )
+          ),
+      ]
+    );
+
+    return Object.fromEntries(assetEntriesByChainId) as Record<
+      keyof AssetsByChainId,
+      (AssetsByChainId[string][number] & {
+        supportedVariants: string[];
+        supportedProviders: Bridge[];
+      })[]
+    >;
+  }, [successfulQueries]);
+
+  console.log(supportedAssetsByChainId);
+
+  const supportedChains = useMemo(() => {
```

Comment:
Thoughts on extracting chains from `supportedAssetsByChainId` instead? The data already has some processing.

Also, we could return that as a tuple from that single `useMemo`

Reply:
I separated them to maintain separation of concerns. I want to avoid complicating supportedAssetsByChainId too much by adding a tuple